src/videoencodingmonitor/VideoEncodingMonitor.ts

```diff
@@ -0,0 +1,173 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import AudioVideoControllerState from '../audiovideocontroller/AudioVideoControllerState';
+import ClientMetricReport from '../clientmetricreport/ClientMetricReport';
+import VideoCodecCapability from '../sdp/VideoCodecCapability';
+
+/**
+ * [[VideoEncodingMonitor]] monitors video encoding issues and degrade video codec when necessary.
+ */
+export default class VideoEncodingMonitor {
+  private continuousHighEncodeCpuCnt: number = 0;
+  private continuousHwEncodeFailureCnt: number = 0;
+  static readonly CONTINUOUS_HIGH_ENCODE_CPU_THRESHOLD = 10;
+  static readonly CONTINUOUS_HW_ENCODE_FAILURE_THRESHOLD = 5;
+  static readonly HIGH_ENCODE_CPU_MS_THRESHOLD: number = 500;
+
+  /**
+   * @param context Take context from monitor task
+   */
+  constructor(private context: AudioVideoControllerState) {}
+
+  /**
+   * Check video encoding metrics for high CPU usage with software encoder
+   * software encoder and degrade video codec when necessary.
+   */
+  private checkHighEncodeCpu(videoMetricReport: { [id: string]: { [id: string]: {} } }): void {
```

Comment:
nit: we can probably split this up into injectable components `VideoEncodingIssueDetector` and `VideoEncodingIssueObserver` or something like that just so we make it easy for builders to either adjust either of those logics seperately without needed to copy paste other code.

Reply:
Splited components into two injectable `ConnectionHealthPolicy`s